### PR TITLE
use workspace package to locate `lib/`

### PR DESCRIPTION
**Pull request by pq**
Fixes: #1786

Should also make `prefer_relative_imports` and `public_member_api_docs` google3-friendly.

Bonus: makes `avoid_renaming_method_parameters` way more performant.

Before:

image

After:

image

🏅 🏅 🏅 

/cc @bwilkerson @scheglov 

/fyi @a14n

### CODE REVIEW

File: lib/src/ast.dart [hunk 1]
```diff
@@ -151,6 +121,14 @@ bool isHashCode(ClassMember element) =>
     (element is FieldDeclaration &&
         getFieldIdentifier(element, 'hashCode') != null);
 
+/// Return true if this compilation unit [node] is declared within the given
+/// [package]'s `lib/` directory.
+bool isInLibDir(CompilationUnit node, WorkspacePackage package) {
+  final libDir = path.join(package.root, 'lib');
+  final cuPath = node.declaredElement.library?.source?.fullName;
+  return cuPath?.startsWith(libDir) == true;
```
Comment (bwilkerson) [inline lib/src/ast.dart]: Consider using `path.isWithin`.

Comment (pq) [reply lib/src/ast.dart]: Cool.  Will-do!

File: lib/src/rules/avoid_renaming_method_parameters.dart [hunk 2]
```diff
@@ -55,18 +55,29 @@ class AvoidRenamingMethodParameters extends LintRule implements NodeLintRule {
...
+    registry.addCompilationUnit(this, visitor);
   }
 }
 
 class _Visitor extends SimpleAstVisitor<void> {
   final LintRule rule;
+  final LinterContext context;
 
-  _Visitor(this.rule);
+  bool isInLib;
```
Comment (srawlins) [inline lib/src/rules/avoid_renaming_method_parameters.dart]: I think this warrants a comment; just saying that it specifically states whether the _compilation unit_ is in [context]'s package.

File: lib/src/ast.dart [hunk 3]
```diff
@@ -151,6 +121,15 @@ bool isHashCode(ClassMember element) =>
     (element is FieldDeclaration &&
         getFieldIdentifier(element, 'hashCode') != null);
 
+/// Return true if this compilation unit [node] is declared within the given
+/// [package]'s `lib/` directory.
```
Comment (srawlins) [inline lib/src/ast.dart]: To differentiate between, say, "public interface" files immediately in lib/, and source files in, e.g. lib/src/, maybe say "declared within the given package's `lib/` directory _tree_."

Comment (pq) [reply lib/src/ast.dart]: Good idea.

File: lib/src/rules/avoid_renaming_method_parameters.dart (hunk 2 above)
Comment (pq) [reply lib/src/rules/avoid_renaming_method_parameters.dart]: Sure thing.  Will do.

**Comment by pq** [comment]
Thanks for the feedback @srawlins!

**Pull request merged**